internal/transform/contract_data.go

```diff
@@ -0,0 +1,503 @@
+
+var (
+	// https://github.com/stellar/rs-soroban-env/blob/v0.0.16/soroban-env-host/src/native_contract/token/public_types.rs#L22
+	nativeAssetSym = xdr.ScSymbol("Native")
+	// these are storage DataKey enum
+	// https://github.com/stellar/rs-soroban-env/blob/v0.0.16/soroban-env-host/src/native_contract/token/storage_types.rs#L23
+	balanceMetadataSym = xdr.ScSymbol("Balance")
+	metadataSym        = xdr.ScSymbol("METADATA")
+	metadataNameSym    = xdr.ScSymbol("name")
+	metadataSymbolSym  = xdr.ScSymbol("symbol")
+	adminSym           = xdr.ScSymbol("Admin")
+	issuerSym          = xdr.ScSymbol("issuer")
+	assetCodeSym       = xdr.ScSymbol("asset_code")
+	alphaNum4Sym       = xdr.ScSymbol("AlphaNum4")
+	alphaNum12Sym      = xdr.ScSymbol("AlphaNum12")
+	decimalSym         = xdr.ScSymbol("decimal")
+	assetInfoSym       = xdr.ScSymbol("AssetInfo")
+	decimalVal         = xdr.Uint32(scDecimalPrecision)
+	assetInfoVec       = &xdr.ScVec{
+		xdr.ScVal{
+			Type: xdr.ScValTypeScvSymbol,
+			Sym:  &assetInfoSym,
+		},
+	}
+	assetInfoKey = xdr.ScVal{
+		Type: xdr.ScValTypeScvVec,
+		Vec:  &assetInfoVec,
+	}
+)
+
+type AssetFromContractDataFunc func(ledgerEntry xdr.LedgerEntry, passphrase string) (string, string)
+type ContractBalanceFromContractDataFunc func(ledgerEntry xdr.LedgerEntry, passphrase string) ([32]byte, *big.Int, bool)
+
+type TransformContractDataStruct struct {
+	AssetFromContractData           AssetFromContractDataFunc
+	ContractBalanceFromContractData ContractBalanceFromContractDataFunc
+}
+
+func NewTransformContractDataStruct(assetfrom AssetFromContractDataFunc, contractBalance ContractBalanceFromContractDataFunc) *TransformContractDataStruct {
+	return &TransformContractDataStruct{
+		AssetFromContractData:           assetfrom,
+		ContractBalanceFromContractData: contractBalance,
+	}
+}
+
+// TransformContractData converts a contract data ledger change entry into a form suitable for BigQuery
+func (t *TransformContractDataStruct) TransformContractData(ledgerChange ingest.Change, passphrase string) (ContractDataOutput, error) {
+	ledgerEntry, changeType, outputDeleted, err := utils.ExtractEntryFromChange(ledgerChange)
+	if err != nil {
+		return ContractDataOutput{}, err
+	}
+
+	contractData, ok := ledgerEntry.Data.GetContractData()
+	if !ok {
+		return ContractDataOutput{}, fmt.Errorf("Could not extract contract data from ledger entry; actual type is %s", ledgerEntry.Data.Type)
+	}
+
+	// LedgerEntryChange must contain a contract data change to be parsed, otherwise skip
+	if ledgerEntry.Data.Type != xdr.LedgerEntryTypeContractData {
+		return ContractDataOutput{}, nil
+	}
+
+	contractDataAssetCode, contractDataAssetIssuer := t.AssetFromContractData(ledgerEntry, passphrase)
+
+	contractDataBalanceHolder, contractDataBalance, _ := t.ContractBalanceFromContractData(ledgerEntry, passphrase)
+
+	if contractData.Key.Type.String() == "ScValTypeScvLedgerKeyNonce" {
+		// Is a nonce and should be discarded
+		return ContractDataOutput{}, nil
+	}
```

Comment:
should this be moved up in the function so that we're continuing out of none key types faster?